File: runtime/server-kt/agent-core/src/main/kotlin/com/garive/runtime/server/agent/ContextSurface.kt
package com.garive.runtime.server.agent

import com.garive.runtime.server.llm.MediaKind
import com.garive.runtime.server.llm.ModelInputContent
import com.garive.runtime.server.llm.ModelInputItem

enum class ContextPurpose { INFERENCE, GOVERNANCE, TOOL_PREPARATION, SUMMARIZATION }

enum class CandidateKind {
    INSTRUCTION,
    USER_INPUT,
    MODEL_OUTPUT,
    TOOL_OBSERVATION,
    APPROVAL,
    SUMMARY,
    SYSTEM_NOTICE,
}

enum class Retention { REQUIRED, OPTIONAL }

sealed interface Visibility {
    data object Visible : Visibility
    data object Redacted : Visibility
    data class Purposes(val purposes: Set<ContextPurpose>) : Visibility
}

data class FactRef(val sessionId: String, val position: ULong)

data class ContextCandidate(
    val factRef: FactRef,
    val kind: CandidateKind,
    val retention: Retention,
    val visibility: Visibility,
    val items: List<ModelInputItem>,
)

data class ContextRequest(
    val sessionId: String,
    val turnId: String,
    val purpose: ContextPurpose,
    val afterPosition: ULong?,
    val throughPosition: ULong,
    val maxItems: Int,
    val maxUtf8Bytes: Int,
)

sealed interface ContextItem {
    data class Input(val factRef: FactRef, val item: ModelInputItem) : ContextItem
    data class RedactedItem(val factRef: FactRef) : ContextItem
}

data class ContextSurface(
    val purpose: ContextPurpose,
    val fromPosition: ULong,
    val throughPosition: ULong,
    val items: List<ContextItem>,
    val retainedRefs: List<FactRef>,
    val droppedRefs: List<FactRef>,
    val filteredRefs: List<FactRef>,
    val itemCount: Int,
    val utf8Bytes: Int,
)

sealed interface ContextDerivationResult {
    data class Success(val surface: ContextSurface) : ContextDerivationResult
    data class Failure(val error: ContextDerivationError) : ContextDerivationResult
}

sealed class ContextDerivationError(val code: String) {
    data object InvalidRequest : ContextDerivationError("invalid-request")
    data object SessionMismatch : ContextDerivationError("session-mismatch")
    data object PositionBeyondSurface : ContextDerivationError("position-beyond-surface")
    data object NonIncreasingPosition : ContextDerivationError("non-increasing-position")
    data object DuplicateReference : ContextDerivationError("duplicate-reference")
    data object EmptyRequiredContent : ContextDerivationError("empty-required-content")
    data object InvalidVisibility : ContextDerivationError("invalid-visibility")
    data object BudgetOverflow : ContextDerivationError("budget-overflow")
    data class RequiredFactsExceedBudget(val itemCount: Int, val utf8Bytes: Int) :
        ContextDerivationError("required-facts-exceed-budget")
}

private data class Eligible(
    val candidate: ContextCandidate,
    val itemCount: Int,
    val utf8Bytes: Int,
    val redacted: Boolean,
)

fun deriveContext(
    request: ContextRequest,
    candidates: List<ContextCandidate>,
): ContextDerivationResult {
    if (request.sessionId.isEmpty() || request.turnId.isEmpty() ||
        request.throughPosition == 0uL || request.maxItems <= 0 || request.maxUtf8Bytes <= 0 ||
        request.afterPosition?.let { it >= request.throughPosition || it == ULong.MAX_VALUE } == true
    ) {
        return ContextDerivationResult.Failure(ContextDerivationError.InvalidRequest)
    }
    val fromPosition = (request.afterPosition ?: 0uL) + 1uL
    val filtered = mutableListOf<FactRef>()
    val eligible = mutableListOf<Eligible>()
    var lastPosition: ULong? = null

    for (candidate in candidates) {
        val validation = validateCandidate(request, candidate, lastPosition)
        if (validation != null) return ContextDerivationResult.Failure(validation)
        lastPosition = candidate.factRef.position
        val visible = when (val visibility = candidate.visibility) {
            Visibility.Visible, Visibility.Redacted -> true
            is Visibility.Purposes -> {
                if (visibility.purposes.isEmpty()) {
                    return ContextDerivationResult.Failure(ContextDerivationError.InvalidVisibility)
                }
                request.purpose in visibility.purposes
            }
        }
        if (candidate.factRef.position <= (request.afterPosition ?: 0uL) || !visible) {
            filtered += candidate.factRef
            continue
        }
        val redacted = candidate.visibility == Visibility.Redacted
        val cost = if (redacted) 1 to 0 else candidateCost(candidate.items)
            ?: return ContextDerivationResult.Failure(ContextDerivationError.BudgetOverflow)
        if (candidate.retention == Retention.REQUIRED && !redacted && (cost.first == 0 || cost.second == 0)) {
            return ContextDerivationResult.Failure(ContextDerivationError.EmptyRequiredContent)
        }
        eligible += Eligible(candidate, cost.first, cost.second, redacted)
    }

    var itemCount = 0
    var utf8Bytes = 0
    val retainedPositions = mutableSetOf<ULong>()
    for (value in eligible.filter { it.candidate.retention == Retention.REQUIRED }) {
        itemCount = checkedAdd(itemCount, value.itemCount)
            ?: return ContextDerivationResult.Failure(ContextDerivationError.BudgetOverflow)
        utf8Bytes = checkedAdd(utf8Bytes, value.utf8Bytes)
            ?: return ContextDerivationResult.Failure(ContextDerivationError.BudgetOverflow)
        retainedPositions += value.candidate.factRef.position
    }
    if (itemCount > request.maxItems || utf8Bytes > request.maxUtf8Bytes) {
        return ContextDerivationResult.Failure(
            ContextDerivationError.RequiredFactsExceedBudget(itemCount, utf8Bytes),
        )
    }
    for (value in eligible.asReversed()) {
        if (value.candidate.retention == Retention.REQUIRED) continue
        val nextItems = checkedAdd(itemCount, value.itemCount)
            ?: return ContextDerivationResult.Failure(ContextDerivationError.BudgetOverflow)
        val nextBytes = checkedAdd(utf8Bytes, value.utf8Bytes)
            ?: return ContextDerivationResult.Failure(ContextDerivationError.BudgetOverflow)
        if (nextItems <= request.maxItems && nextBytes <= request.maxUtf8Bytes) {
            itemCount = nextItems
            utf8Bytes = nextBytes
            retainedPositions += value.candidate.factRef.position
        }
    }

    val items = mutableListOf<ContextItem>()
    val retained = mutableListOf<FactRef>()
    val dropped = mutableListOf<FactRef>()
    for (value in eligible) {
        if (value.candidate.factRef.position !in retainedPositions) {
            dropped += value.candidate.factRef
            continue
        }
        retained += value.candidate.factRef
        if (value.redacted) {
            items += ContextItem.RedactedItem(value.candidate.factRef)
        } else {
            items += value.candidate.items.map { ContextItem.Input(value.candidate.factRef, it) }
        }
    }
    return ContextDerivationResult.Success(
        ContextSurface(
            request.purpose,
            fromPosition,
            request.throughPosition,
            items,
            retained,
            dropped,
            filtered,
            itemCount,
            utf8Bytes,
        ),
    )
}

private fun validateCandidate(
    request: ContextRequest,
    candidate: ContextCandidate,
    lastPosition: ULong?,
): ContextDerivationError? {
    if (candidate.factRef.sessionId != request.sessionId) return ContextDerivationError.SessionMismatch
    if (candidate.factRef.position == 0uL || candidate.factRef.position > request.throughPosition) {
        return ContextDerivationError.PositionBeyondSurface
    }
    if (candidate.factRef.position == lastPosition) return ContextDerivationError.DuplicateReference
    if (lastPosition != null && candidate.factRef.position < lastPosition) {
        return ContextDerivationError.NonIncreasingPosition
    }
    if (candidate.retention == Retention.REQUIRED && candidate.items.isEmpty()) {
        return ContextDerivationError.EmptyRequiredContent
    }
    return null
}

private fun candidateCost(items: List<ModelInputItem>): Pair<Int, Int>? {
    var bytes = 0
    for (item in items) bytes = checkedAdd(bytes, itemUtf8Bytes(item) ?: return null) ?: return null
    return items.size to bytes
}

private fun itemUtf8Bytes(item: ModelInputItem): Int? {
    val strings = when (item) {
        is ModelInputItem.Message -> item.content.flatMap { content ->
            when (content) {
                is ModelInputContent.Text -> listOf(content.text)
                is ModelInputContent.MediaReference -> buildList {
                    add(content.reference)
                    add(content.mediaType)
                    val mediaKind = content.mediaKind
                    if (mediaKind is MediaKind.Other) add(mediaKind.name)
                }
            }
        }
        is ModelInputItem.ToolObservation -> listOf(item.modelCallId, item.resultJson)
        is ModelInputItem.ReasoningReference -> listOf(item.reference)
    }
    var total = 0
    for (value in strings) total = checkedAdd(total, value.encodeToByteArray().size) ?: return null
    return total
}

private fun checkedAdd(left: Int, right: Int): Int? =
    if (right < 0 || Int.MAX_VALUE - left < right) null else left + right

File: engine/core/tests/context_surface.rs
use std::{collections::BTreeSet, fs, path::PathBuf};

use garive_core::{
    derive_context, CandidateKind, ContextCandidate, ContextDerivationError, ContextItem,
    ContextPurpose, ContextRequest, FactRef, Retention, Visibility,
};
use garive_llm::{MediaKind, ModelInputContent, ModelInputItem, ModelRole};
use serde_json::Value;

fn fixture() -> Value {
    let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../spec/fixtures/agent/context-surface.json");
    serde_json::from_str(&fs::read_to_string(path).unwrap()).unwrap()
}

fn purpose(value: &str) -> ContextPurpose {
    match value {
        "inference" => ContextPurpose::Inference,
        "governance" => ContextPurpose::Governance,
        "tool-preparation" => ContextPurpose::ToolPreparation,
        "summarization" => ContextPurpose::Summarization,
        other => panic!("unknown purpose: {other}"),
    }
}

fn request(value: &Value) -> ContextRequest {
    ContextRequest {
        session_id: "session-1".into(),
        turn_id: "turn-1".into(),
        purpose: purpose(value["purpose"].as_str().unwrap()),
        after_position: value["after"].as_u64(),
        through_position: value["through"].as_u64().unwrap(),
        max_items: value["max_items"].as_u64().unwrap() as usize,
        max_utf8_bytes: value["max_bytes"].as_u64().unwrap() as usize,
    }
}

fn visibility(value: &str) -> Visibility {
    match value {
        "visible" => Visibility::Visible,
        "redacted" => Visibility::Redacted,
        limited if limited.starts_with("purpose:") => {
            Visibility::Purposes(BTreeSet::from([purpose(
                limited.strip_prefix("purpose:").unwrap(),
            )]))
        }
        other => panic!("unknown visibility: {other}"),
    }
}

fn kind(value: Option<&str>) -> CandidateKind {
    match value.unwrap_or("user-input") {
        "instruction" => CandidateKind::Instruction,
        "user-input" => CandidateKind::UserInput,
        "model-output" => CandidateKind::ModelOutput,
        "tool-observation" => CandidateKind::ToolObservation,
        "approval" => CandidateKind::Approval,
        "summary" => CandidateKind::Summary,
        "system-notice" => CandidateKind::SystemNotice,
        other => panic!("unknown candidate kind: {other}"),
    }
}

fn candidates(values: &Value) -> Vec<ContextCandidate> {
    values
        .as_array()
        .unwrap()
        .iter()
        .map(|value| ContextCandidate {
            fact_ref: FactRef {
                session_id: value["session"].as_str().unwrap_or("session-1").into(),
                position: value["position"].as_u64().unwrap(),
            },
            kind: kind(value["kind"].as_str()),
            retention: match value["retention"].as_str().unwrap() {
                "required" => Retention::Required,
                "optional" => Retention::Optional,
                other => panic!("unknown retention: {other}"),
            },
            visibility: visibility(value["visibility"].as_str().unwrap()),
            items: value["items"]
                .as_array()
                .unwrap()
                .iter()
                .map(|item| ModelInputItem::Message {
                    role: ModelRole::User,
                    content: vec![ModelInputContent::Text(item.as_str().unwrap().into())],
                })
                .collect(),
        })
        .collect()
}

fn positions(value: &Value) -> Vec<u64> {
    value
        .as_array()
        .unwrap()
        .iter()
        .map(|v| v.as_u64().unwrap())
        .collect()
}

fn rendered_items(items: &[ContextItem]) -> Vec<String> {
    items
        .iter()
        .map(|item| match item {
            ContextItem::Input {
                item: ModelInputItem::Message { content, .. },
                ..
            } => match &content[0] {
                ModelInputContent::Text(text) => format!("text:{text}"),
                _ => panic!("unexpected media fixture"),
            },
            ContextItem::RedactedItem { .. } => "redacted".into(),
            _ => panic!("unexpected fixture item"),
        })
        .collect()
}

#[test]
fn rust_consumes_every_context_case() {
    let document = fixture();
    let cases = document["cases"].as_array().unwrap();
    assert_eq!(cases.len(), 12);
    for case in cases {
        let request = request(&case["request"]);
        let expected = &case["expected"];
        let result = derive_context(&request, &candidates(&case["candidates"]));
        if expected["status"] != "ok" {
            let error = result.unwrap_err();
            assert_eq!(error.code(), expected["status"], "{}", case["name"]);
            continue;
        }
        let surface = result.unwrap();
        assert_eq!(surface.purpose, request.purpose, "{}", case["name"]);
        assert_eq!(
            surface.from_position,
            request.after_position.unwrap_or(0) + 1,
            "{}",
            case["name"]
        );
        assert_eq!(
            surface.through_position, request.through_position,
            "{}",
            case["name"]
        );
        assert_eq!(
            surface
                .retained_refs
                .iter()
                .map(|r| r.position)
                .collect::<Vec<_>>(),
            positions(&expected["retained"])
        );
        assert_eq!(
            surface
                .dropped_refs
                .iter()
                .map(|r| r.position)
                .collect::<Vec<_>>(),
            positions(&expected["dropped"])
        );
        assert_eq!(
            surface
                .filtered_refs
                .iter()
                .map(|r| r.position)
                .collect::<Vec<_>>(),
            positions(&expected["filtered"])
        );
        assert_eq!(
            rendered_items(&surface.items),
            expected["items"]
                .as_array()
                .unwrap()
                .iter()
                .map(|v| v.as_str().unwrap().to_owned())
                .collect::<Vec<_>>()
        );
        assert_eq!(
            surface.item_count,
            expected["item_count"].as_u64().unwrap() as usize
        );
        assert_eq!(
            surface.utf8_bytes,
            expected["bytes"].as_u64().unwrap() as usize
        );
    }
}

#[test]
fn request_and_candidate_boundaries_fail_closed() {
    let base = request(&fixture()["cases"][0]["request"]);
    let invalid_requests = [
        ContextRequest {
            session_id: String::new(),
            ..base.clone()
        },
        ContextRequest {
            turn_id: String::new(),
            ..base.clone()
        },
        ContextRequest {
            through_position: 0,
            ..base.clone()
        },
        ContextRequest {
            max_items: 0,
            ..base.clone()
        },
        ContextRequest {
            max_utf8_bytes: 0,
            ..base.clone()
        },
        ContextRequest {
            after_position: Some(base.through_position),
            ..base.clone()
        },
        ContextRequest {
            after_position: Some(u64::MAX),
            ..base.clone()
        },
    ];
    for request in invalid_requests {
        assert_eq!(
            derive_context(&request, &[]),
            Err(ContextDerivationError::InvalidRequest)
        );
    }

    let candidate =
        |session_id: &str, position: u64, items: Vec<ModelInputItem>| ContextCandidate {
            fact_ref: FactRef {
                session_id: session_id.into(),
                position,
            },
            kind: CandidateKind::Instruction,
            retention: Retention::Required,
            visibility: Visibility::Visible,
            items,
        };
    let text = || {
        vec![ModelInputItem::Message {
            role: ModelRole::System,
            content: vec![ModelInputContent::Text("instruction".into())],
        }]
    };
    assert_eq!(
        derive_context(&base, &[candidate("other", 2, text())]),
        Err(ContextDerivationError::SessionMismatch)
    );
    assert_eq!(
        derive_context(&base, &[candidate("session-1", 0, text())]),
        Err(ContextDerivationError::PositionBeyondSurface)
    );
    assert_eq!(
        derive_context(&base, &[candidate("session-1", 5, text())]),
        Err(ContextDerivationError::PositionBeyondSurface)
    );
    assert_eq!(
        derive_context(&base, &[candidate("session-1", 2, Vec::new())]),
        Err(ContextDerivationError::EmptyRequiredContent)
    );
    let mut empty_visibility = candidate("session-1", 2, text());
    empty_visibility.visibility = Visibility::Purposes(BTreeSet::new());
    assert_eq!(
        derive_context(&base, &[empty_visibility]),
        Err(ContextDerivationError::InvalidVisibility)
    );
}

#[test]
fn every_model_input_payload_field_counts_toward_the_budget() {
    let request = ContextRequest {
        session_id: "session-1".into(),
        turn_id: "turn-1".into(),
        purpose: ContextPurpose::Inference,
        after_position: None,
        through_position: 1,
        max_items: 3,
        max_utf8_bytes: 31,
    };
    let candidate = ContextCandidate {
        fact_ref: FactRef {
            session_id: "session-1".into(),
            position: 1,
        },
        kind: CandidateKind::Instruction,
        retention: Retention::Required,
        visibility: Visibility::Visible,
        items: vec![
            ModelInputItem::Message {
                role: ModelRole::System,
                content: vec![
                    ModelInputContent::Text("a".into()),
                    ModelInputContent::MediaReference {
                        media_kind: MediaKind::Other("custom".into()),
                        reference: "ref".into(),
                        media_type: "image/png".into(),
                    },
                ],
            },
            ModelInputItem::ToolObservation {
                model_call_id: "call".into(),
                result_json: "{}".into(),
            },
            ModelInputItem::ReasoningReference {
                reference: "reason".into(),
            },
        ],
    };
    let surface = derive_context(&request, &[candidate]).unwrap();
    assert_eq!(surface.item_count, 3);
    assert_eq!(surface.utf8_bytes, 31);
}
